File: SenLib/Sen3/FileFixes/t3510_dat.cs
using HyoutaUtils;
using HyoutaUtils.Streams;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SenLib.Sen3.FileFixes {
	class t3510_dat : FileMod {
		public string GetDescription() {
			return "Terminology fixes in Ordis (Duke's Castle Entrance).";
		}

		public IEnumerable<FileModResult> TryApply(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x7d6eddb306de371ful, 0xd7fcfd55b5993016ul, 0x83e18853u));
			if (file == null) {
				return null;
			}

			var bin = file.CopyToMemory();

			// the port city -> the Port City
			bin.Position = 0x4401;
			bin.WriteUInt8(0x50);
			bin.Position = 0x4406;
			bin.WriteUInt8(0x43);

			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_en/t3510.dat", bin) };
		}

		public IEnumerable<FileModResult> TryRevert(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x7d6eddb306de371ful, 0xd7fcfd55b5993016ul, 0x83e18853u));
			if (file == null) {
				return null;
			}
			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_en/t3510.dat", file) };
		}
	}
}

File: native/sen3/FileFixes/t3510_dat.cpp
#include <filesystem>
#include <string_view>
#include <vector>

#include "file_getter.h"

#include "p3a/pack.h"
#include "p3a/structs.h"
#include "sha1.h"

namespace SenLib::Sen3::FileFixes::t3510_dat {
std::string_view GetDescription() {
    return "Terminology fixes in Ordis (Duke's Castle Entrance).";
}

bool TryApply(const SenPatcher::GetCheckedFileCallback& getCheckedFile,
              std::vector<SenPatcher::P3APackFile>& result) {
    try {
        auto file = getCheckedFile(
            "data/scripts/scena/dat_en/t3510.dat",
            64009,
            SenPatcher::SHA1FromHexString("7d6eddb306de371fd7fcfd55b599301683e18853"));
        if (!file) {
            return false;
        }

        auto& bin = file->Data;

        // the port city -> the Port City
        bin[0x4401] = 0x50;
        bin[0x4406] = 0x43;

        result.emplace_back(std::move(bin), file->Filename, SenPatcher::P3ACompressionType::LZ4);

        return true;
    } catch (...) {
        return false;
    }
}
} // namespace SenLib::Sen3::FileFixes::t3510_dat
